### ISSUE TITLE
WebP with alpha image is broken

### ISSUE THREAD

**Issue by neelabo** [dotnet/wpf#1436]
* .NET Core Version: 3.0 Preview7
* Windows version: Windows 10 version 1903
* Does the bug reproduce also in WPF for .NET Framework 4.8?: Yes

**Problem description:**
 
WebP can now be displayed on Windows 10, but when trying to display transparent WebP with WPF, the transparent part becomes broken display.

![sample](https://user-images.githubusercontent.com/16382967/62021615-1a748280-b203-11e9-9d3e-3a63fcc8c1ed.png)

However, BitmapImage using DecodePixelWidth different from the source is displayed correctly.

[I submitted a similar report as an issue with the .NET Framework.](https://developercommunity.visualstudio.com/content/problem/465396/webp-with-alpha-image-is-broken.html)

**Minimal repro:**

* The test WebP image is [here](https://developers.google.com/speed/webp/gallery2).
* `<Image Source="1_webp_a.webp"/>` only.

**Comment by jairbubbles** [comment]
Indeed, same behavior on .NET Framework, the .webp version is broken while the .png version displays correctly.

This code:

```xaml
<UniformGrid Rows="1">
     <Image Source="https://www.gstatic.com/webp/gallery3/1_webp_ll.webp"></Image>
     <Image Source="https://www.gstatic.com/webp/gallery3/1_webp_ll.png"></Image>
</UniformGrid>
```
Will produce:

![image](https://user-images.githubusercontent.com/15875066/69436858-7c24a380-0d42-11ea-8627-58a32dc44497.png)

(Windows version 1903 too)

**Comment by h82258652** [comment]
https://github.com/dotnet/wpf/blob/ac9d1b7a6b0ee7c44fd2875a1174b820b3940619/src/Microsoft.DotNet.Wpf/src/PresentationCore/System/Windows/Media/Imaging/BitmapDecoder.cs
There is no webp decoder. So wpf use the ```UnknownBitmapDecoder``` and it can not handle the alpha in webp format.

**Comment by weltkante** [comment]
> There is no webp decoder. So wpf use the UnknownBitmapDecoder

Thats a contradiction, I assume you meant to say that there's no explicit C#/.NET class for it and (I assume) it just falls back to WIC and lets Windows resolve the decoder. All decoders are native implementations just some have explicit wrappers in .NET - but if you look at their implementation they all do the same as the UnknownBitmapDecoder and forward to base class

**Comment by weltkante** [comment]
@grubioe can we get that reevaluated? If this is a bug in WIC it would be really annoying if its not forwarded to the right Windows team just because the WPF team planned it for "future"

**Comment by jairbubbles** [comment]
The thumbnail appears correctly on Windows explorer. I don't know if it ends up using the same decoder though.

**Comment by weltkante** [comment]
I'll test with native WIC when I get home to determine if this is a WPF bug or WIC bug, for some reason my Windows 10 / 1909 installation at work doesn't have the webp decoder (I get an exception when referencing the image in WPF) and Windows doesn't offer it as optional feature either. I'm pretty sure I've seen it yesterday on my home machine when I happened to scroll through optional features.

**Comment by weltkante** [comment]
The bug is in WPF, WIC can load the image just fine. Test code to load the image via WIC attached.

In fact I also figured out what happens, WPF simply ignores the alpha channel even though WIC reports that the image has an alpha channel. The webp image has color artifacts in transparent places so it looks weird when they aren't transparent.

<details><summary>example code to load image via WIC</summary>

```csharp
private void LoadImage()
{
    // Referencing SharpDX.Direct2D1 nuget package
    using (var factory = new SharpDX.WIC.ImagingFactory2())
    using (var decoder = new SharpDX.WIC.BitmapDecoder(factory, "1_webp_a.webp", SharpDX.WIC.DecodeOptions.CacheOnLoad))
    {
        Debug.Assert(decoder.FrameCount == 1);
        using (var frame = decoder.GetFrame(0))
        {
            Debug.Assert(frame.PixelFormat == SharpDX.WIC.PixelFormat.Format32bppRGBA);
            var size = frame.Size;
            var stride = size.Width * 4;
            var pixels = new byte[stride * size.Height];
            frame.CopyPixels(pixels, stride);

            for (int iy = 0; iy < size.Height; iy++)
            {
                for (int ix = 0; ix < size.Width; ix++)
                {
                    // convert from RGBA to BGRA
                    Swap(ref pixels[iy * stride + ix * 4 + 0], ref pixels[iy * stride + ix * 4 + 2]);

                    // uncomment to reproduce the bug, WPF seems to ignore the alpha channel
                    // which you can simulate by setting alpha to 255 (opaque)
                    //pixels[iy * stride + ix * 4 + 3] = 255;
                }
            }

            var bitmap = new WriteableBitmap(size.Width, size.Height, 96, 96, PixelFormats.Bgra32, null);
            bitmap.Lock();
            try { bitmap.WritePixels(new Int32Rect(0, 0, size.Width, size.Height), pixels, stride, 0); }
            finally { bitmap.Unlock(); }
            ImageControl.Source = bitmap;
        }
    }
}

private static void Swap(ref byte lhs, ref byte rhs)
{
    var tmp = lhs;
    lhs = rhs;
    rhs = tmp;
}
```

</details>

**Comment by vatsan-madhavan** [comment]
@arpitmathur can we bring this one up in our bug triage again for discussion ? 

More than just WebP, I’d like us to think about how we are going to support emerging media formats for which _native_ WIC support is lacking. We may want to find out how Office does this today, for e.g. to see if their approach is reusable for us. 

{It could be that we wait till Jan to dig into this.}

**Comment by jairbubbles** [comment]
Would be nice to have native support for .gif. I usually use https://github.com/XamlAnimatedGif/WpfAnimatedGif for this.

**Comment by icetech233** [comment]
我也发现了 这个 bug ，怎么处理
！？？

**Comment by miloush** [comment]
@icetech233 See https://github.com/dotnet/wpf/issues/1436#issuecomment-562269664 for a workaround. Basically you need to convert the image into a known pixel format. You can let WIC do it:

```C#
var decoder = BitmapDecoder.Create(new Uri("1_webp_a.webp", UriKind.Relative), BitmapCreateOptions.PreservePixelFormat, BitmapCacheOption.None);
var converted = new FormatConvertedBitmap(decoder.Frames[0], PixelFormats.Bgra32, null, 0);

Content = new Image { Source = converted };
```

The WebP decoder returns `GUID_WICPixelFormat32bppRGBA` which is not supported by WPF so it fallbacks to `Default`. I suggested extending the supported pixel formats in #4570

**Comment by oktagon2** [comment]
Based on the example in [learn.microsoft](https://learn.microsoft.com/en-us/dotnet/desktop/wpf/graphics-multimedia/imaging-overview?view=netframeworkdesktop-4.8#displaying-images-in-wpf) I coded the following xaml:

```
<Window x:Class="WpfApplicationForStackOverflow1.MainWindow"
        xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        xmlns:local="clr-namespace:WpfApplicationForStackOverflow1"
        mc:Ignorable="d"
        Title="MainWindow" Height="450" Width="800"
        Background="Aquamarine">
    <StackPanel>
        <Image Name="Image1" Width="400">
            <Image.Source>
                <BitmapImage DecodePixelWidth="997"
                             UriSource="https://www.gstatic.com/webp/gallery3/1_webp_ll.webp" />
            </Image.Source> 
        </Image>
    </StackPanel>
</Window>
```
It works. The DecodePixelWidth seems to have a healthy effect. Unfortunately not for every value. I prefer prime numbers like 997.

**Comment by neelabo** [comment]
[JPEG XL Image Extension](https://apps.microsoft.com/detail/9MZPRTH5C0TB) also seems to have issues displaying alpha images correctly.

#### Sample Images
- [dice.zip](https://github.com/user-attachments/files/22844441/dice.zip) from https://jpegxl.info/resources/jpeg-xl-test-page.html

**Comment by lindexi** [comment]
@neelabo Could you attach the test image files?

**Comment by neelabo** [comment]
@lindexi Since I couldn't attach the .jxl file directly, I've attached dice.zip, a compressed file, to the previous comment.
